Include thruster disturbances

gabrielfpacheco opened the pull request:
Add disturbances sampled from gaussian distributions for each thruster effort in order to emulate input disturbances to the system. 

Test procedure:
* Checkout this branch
* Configure each thruster disturbance parameters (mu, sigma) in the sdf file 
* Build the solution and run the simulation 
* Use the move_to_position or other action to make the robot move
* Inspect on plot2D the thruster efforts output ports in the gazebo_thruster element
* Check whether the robot was able to achieve the goal

Commit:
add vscode folder to gitignore

Commit:
add thruster efforts in simulation

Commit:
add thruster disturbance in simulation

Commit:
remove unnecessary property

Commit:
improving comments on functions descriptions

doudou commented:
Hey. This is a common package, not at all owned (or private to) 13.

I think that what you want to do is make your private pull request depend on this one, not the other way around (this does not depend on the bundle, does it really ?)

doudou commented on src/Utils.cpp:
```diff
@@ -0,0 +1,22 @@
+
+namespace gazebo_thruster
+{
+    double sampleNormal(const double& mu, const double& sigma)
```
Initializing a RNG is pretty expensive ... get rid of sampleNormal, and initialize the RNG directly as an attribute of the plugin class.

gabrielfpacheco replied:
Makes sense, thanks. I'll do that.

doudou commented on src/GazeboThruster.cpp:
```diff
@@ -140,14 +146,17 @@ void GazeboThruster::readInput(ThrustersMSG const& thrustersMSG)
-                thruster->effort = updateEffort(thrusterCMD);
+                disturbance = sampleNormal(thruster->disturbance.mu, 
+                    thruster->disturbance.sigma);
+                thruster->effort = updateEffort(thrusterCMD, disturbance);
```
Is this really what you want ?

I would have applied the disturbance at the simulation tick, not at the input.

gabrielfpacheco replied:
> Hey. This is a common package, not at all owned (or private to) 13.
> 
> I think that what you want to do is make your private pull request depend on this one, not the other way around (this does not depend on the bundle, does it really ?)

Sorry, @doudou. This was my bad, I have accidentally chosen the wrong target to perform the merge, it was intended to be done on 13's fork of this common package. And you are right, the dependency works the other way around, thanks.

gabrielfpacheco replied:
> Is this really what you want ?
> 
> I would have applied the disturbance at the simulation tick, not at the input.

The idea is really to add a disturbance on the thruster effort that will be transmitted to the robot so that the controllers robustness could be assessed. This kind of disturbance could physically represent unexpected behaviors on the mechanic assembly for example. I am not sure that I understood how to apply the disturbance on the simulation tick.

renan028 replied:
> I would have applied the disturbance at the simulation tick, not at the input.

I think he meant that he would put it at the "updateHook" at the orogen Task, or at the "updateBegin" at the library.

renan028 replied:
> > Hey. This is a common package, not at all owned (or private to) 13.
> > I think that what you want to do is make your private pull request depend on this one, not the other way around (this does not depend on the bundle, does it really ?)
> 
> Sorry, @doudou. This was my bad, I have accidentally chosen the wrong target to perform the merge, it was intended to be done on 13's fork of this common package. And you are right, the dependency works the other way around, thanks.

Thanks, Sylvain. Gabriel, you can click Edit and fix the PR to target the fork.

doudou replied:
> > Hey. This is a common package, not at all owned (or private to) 13.
> > I think that what you want to do is make your private pull request depend on this one, not the other way around (this does not depend on the bundle, does it really ?)
> 
> Sorry, @doudou. This was my bad, I have accidentally chosen the wrong target to perform the merge, it was intended to be done on 13's fork of this common package. And you are right, the dependency works the other way around, thanks.

Suit yourself. But I like the idea of integrating that in the common package ... Your choice.

gabrielfpacheco replied:
Cool, let's integrate that on the common package.

doudou reviewed (commented):
Apart from the code comments, I have one general issue: I doubt that a Gaussian noise really makes sense ... My gut feeling is that a form of random walk could make more sense ... what do you think ?

renan028 commented:
> Apart from the code comments, I have one general issue: I doubt that a Gaussian noise really makes sense ... My gut feeling is that a form of random walk could make more sense ... what do you think ?

I think disturbance and noise are different concepts. One can add a White noise (normal distribution) or Brownian noise (random walk) at the final output.
However, disturbances are usually modeled for an Observer (in a control loop). In DC motors, torque disturbances are usually modeled as: 
t_d = K*i_a - J_m * alpha - B_m * omega
where i_a is the current, alpha angular acceleration, omega angular velocity and J_m and B_m are related to the inertial moments. The idea is that when you are on high speed, one can not "feel" disturbances due to motor inertia.

Some papers:
https://scholar.uwindsor.ca/cgi/viewcontent.cgi?article=1116&context=etd 
https://core.ac.uk/download/pdf/72772437.pdf

doudou commented on src/GazeboThruster.cpp:
```diff
@@ -177,7 +177,7 @@ void GazeboThruster::readCompensatedMass(CompMassMSG const& compMassMSG)
 }
 
 double GazeboThruster::updateEffort(gazebo_thruster::msgs::Thruster thrusterCMD,
-    double disturbance)
```
Well. Nope. Don't pass plain doubles as const reference.

doudou commented:
> > Apart from the code comments, I have one general issue: I doubt that a Gaussian noise really makes sense ... My gut feeling is that a form of random walk could make more sense ... what do you think ?
> 
> I think disturbance and noise are different concepts. One can add a White noise (normal distribution) or Brownian noise (random walk) at the final output.
> However, disturbances are usually modeled for an Observer (in a control loop). In DC motors, torque disturbances are usually modeled as:
> t_d = K*i_a - J_m * alpha - B_m * omega
> where i_a is the current, alpha angular acceleration, omega angular velocity and J_m and B_m are related to the inertial moments. The idea is that when you are on high speed, one can not "feel" disturbances due to motor inertia.
> 
> Some papers:
> https://scholar.uwindsor.ca/cgi/viewcontent.cgi?article=1116&context=etd
> https://core.ac.uk/download/pdf/72772437.pdf

Well .. so we agree ?

Commit:
refactor to improve performance

This commit deals with the following:

* The RNG for each thruster is now generated once the thruster is
loaded, only the sampling over the distribution is done at every tick.
* The disturbance is added just after reading the thrusterCMD comming
from the control package. This is not done at udateBegin to avoid the
call in moments when the no thruster message is being sent to the
simulation.

Commit:
update cmakelist and remove logging

gabrielfpacheco commented:
> > > Apart from the code comments, I have one general issue: I doubt that a Gaussian noise really makes sense ... My gut feeling is that a form of random walk could make more sense ... what do you think ?
> > 
> > 
> > I think disturbance and noise are different concepts. One can add a White noise (normal distribution) or Brownian noise (random walk) at the final output.
> > However, disturbances are usually modeled for an Observer (in a control loop). In DC motors, torque disturbances are usually modeled as:
> > t_d = K*i_a - J_m * alpha - B_m * omega
> > where i_a is the current, alpha angular acceleration, omega angular velocity and J_m and B_m are related to the inertial moments. The idea is that when you are on high speed, one can not "feel" disturbances due to motor inertia.
> > Some papers:
> > https://scholar.uwindsor.ca/cgi/viewcontent.cgi?article=1116&context=etd
> > https://core.ac.uk/download/pdf/72772437.pdf
> 
> Well .. so we agree ?

gabrielfpacheco commented:
In our case, we are not modeling the disturbance for the moment but we are trying to evaluate how a constant disturbance (that could be caused by a load for example) would be handled by the controllers. For me, sampling this value from a normal distribution is just a practical way of getting a value centered at the desired vaue. This could also be done using a randon walk, but I am not sure if this would affect much our analysis.

doudou commented:
Mmmm ... I really don't think you'll achieve much in the way of "evaluating how a constant disturbance" would be handled by the controllers.

I believe that you should make a simple model of what you want to model and implement that. It will be minimally more complicated, but allow you to truly test what you want to test (or, at least, much closer to what you will be getting with this implementation)

With this implementation, the "disturbance" will change every time you send a command. Which is usually still much higher frequency that what a mechanically-induced error would generate.

gabrielfpacheco commented:
Don't you think that vibrations could generate such high-frequency behavior? 

The idea using a normal distribution was to capture both a constant disturbance - such as a payload in the thruster - and high-frequency disturbances - such as the ones induced by mechanic vibrations or by current measurement HF noise.

In that way, if only the constant disturbance were to be tested, it suffices to set sigma = 0. From a control perspective, a simple proportional controller is not able to reject such disturbance; integral action is needed and its value can be tuned according to the required disturbance rejection performance. On the other hand, if only high-frequency disturbances were to be tested, one could set mu = 0, but it is true that even the robot's dynamics acts as a filter in such case.

Commit:
handle non-existent disturbanceElement

renan028 committed:
fix verifying if there is disturbance element

doudou commented:
> Don't you think that vibrations could generate such high-frequency behavior?

I honestly don't know. But I very much doubt its generic applicability (i.e. I'm reverting my previous comment, thinking that as-is this aren't worth being included in the plugin)

Beyond the traditional friction effects, something you might rather want to consider if you are controlling your thrusters in rotational velocity is the speed-to-effort conversion which is quadratic. With ramping effects, it would probably be a more interesting target for something the control loop will have to handle.

It all depends on your actual motor/driver combination of course. If you're intent to go on with this simple model, I'd advise creating a separate component in front of gazebo and tune the error model to match whatever you're building.

gabrielfpacheco commented:
> > Don't you think that vibrations could generate such high-frequency behavior?
> 
> I honestly don't know. But I very much doubt its generic applicability (i.e. I'm reverting my previous comment, thinking that as-is this aren't worth being included in the plugin)
> 
> Beyond the traditional friction effects, something you might rather want to consider if you are controlling your thrusters in rotational velocity is the speed-to-effort conversion which is quadratic. With ramping effects, it would probably be a more interesting target for something the control loop will have to handle.
> 
> It all depends on your actual motor/driver combination of course. If you're intent to go on with this simple model, I'd advise creating a separate component in front of gazebo and tune the error model to match whatever you're building.

In that case, I will close this PR for the time being. Thank you for your feedback, Sylvain, it helped a lot. I was indeed already developing the speed-to-effort relationship in the plugin and have also considered a deadband where low rotational speeds lead to no thrust in the output, which has been insightful to analyze in simulation.

gabrielfpacheco commented:
PR closed since the modifications are application-specific.
